service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/CatalogHandlerUtils.java

```diff
@@ -433,39 +461,233 @@ protected TableMetadata commit(TableOperations ops, UpdateTableRequest request)
+    for (UpdateRequirement requirement : request.requirements()) {
+      if (requirement instanceof UpdateRequirement.AssertRefSnapshotID) {
+        ++total;
+        assertRefSnapshotID = (UpdateRequirement.AssertRefSnapshotID) requirement;
+      }
+    }
+
+    // if > 1 assertion for refs, then it's not safe to roll back, make this Noop.
+    return total != 1 ? null : assertRefSnapshotID;
+  }
+
+  private List<MetadataUpdate> generateUpdatesToRemoveNoopSnapshot(
+      TableMetadata base, long expectedCurrentSnapshotId, String updateRefName) {
+    // find the all the snapshots we want to retain which are not the part of current branch.
+    Set<Long> idsToRetain = Sets.newHashSet();
+    for (Map.Entry<String, SnapshotRef> ref : base.refs().entrySet()) {
+      String refName = ref.getKey();
+      SnapshotRef snapshotRef = ref.getValue();
+      if (refName.equals(updateRefName)) {
+        continue;
+      }
+      idsToRetain.add(ref.getValue().snapshotId());
+      // Always check the ancestry for both branch and tags
+      // mostly for case where a branch was created and then was dropped
+      // then a tag was created and then rollback happened post that tag
+      // was dropped and branch was re-created on it.
+      for (Snapshot ancestor : SnapshotUtil.ancestorsOf(snapshotRef.snapshotId(), base::snapshot)) {
+        idsToRetain.add(ancestor.snapshotId());
+      }
+    }
+
+    List<MetadataUpdate> updateToRemoveSnapshot = new ArrayList<>();
+    Long snapshotId = base.ref(updateRefName).snapshotId(); // current tip of the given branch
+    // ensure this branch has the latest sequence number.
+    long expectedSequenceNumber = base.lastSequenceNumber();
+    Set<Long> snapshotsToRemove = new LinkedHashSet<>();
+    while (snapshotId != null && !Objects.equals(snapshotId, expectedCurrentSnapshotId)) {
+      Snapshot snap = base.snapshot(snapshotId);
+      // catch un-expected state the commit sequence number are
+      // not continuous can happen for a table with multiple branches.
+      if (expectedSequenceNumber != snap.sequenceNumber()) {
+        LOGGER.debug(
+            "Giving up rolling back table {} to snapshot {}, Sequence Number are not continuous from {}",
+            base.uuid(),
+            snapshotId,
+            expectedSequenceNumber);
+        break;
+      }
+      if (!isRollbackSnapshot(snap) || idsToRetain.contains(snapshotId)) {
+        // Either encountered a non no-op snapshot or the snapshot is being referenced by any other
+        // reference either by branch or a tag.
+        LOGGER.debug(
+            "Giving up rolling back table {} to snapshot {}, snapshot to be removed referenced by another branch or tag ancestor",
+            base.uuid(),
+            snapshotId);
+        break;
+      }
+      snapshotsToRemove.add(snap.snapshotId());
+      snapshotId = snap.parentId();
+      // we need continuous sequence number to correctly rollback
+      expectedSequenceNumber--;
```

Comment:
I'm not sure this assumption holds :thinking: What if we have two branches, `main` and `dev`, that get commits in an alternating manner, but have only one (very old) commit in common? If we're rolling back commits from `main` they will not have contiguous sequence numbers, but logically they can be rolled back. WDYT?

Do we even need to worry about `expectedSequenceNumber` after we validate the first snapshot to roll back?

Reply:
Its just that it will miss the opportunity to rollback, i think its fine from correctness POV, but i understand we can handle it, let me see if i can add that check !